### DISCUSSION TITLE
Library Editor Features

### DISCUSSION THREAD

**Discussion by x87** [sannybuilder/dev#260] [Team Posts]
https://library.sannybuilder.com/

![1](https://user-images.githubusercontent.com/5698288/113786631-8234dd80-9707-11eb-8972-3cdb3b5e0975.PNG)

**Comment by x87** [comment]
### List of supported games
![02](https://user-images.githubusercontent.com/5698288/113786680-a2fd3300-9707-11eb-8481-b3918950ed5b.PNG)

Displays a series of icons, one per each game where this command is supported. If the command is not supported (`is_unsupported: true`) then no game icon is displayed. Hovering an icon shows a hint text.

Icons could be semi-transparent or have an orange border line:
![03](https://user-images.githubusercontent.com/5698288/113787280-b957be80-9708-11eb-8682-e6da1ba7fe61.PNG)

* the orange border signals that the game supports this command, but the number of parameters is different.
* a transparent icon is displayed for games where the command is a nop.

**Comment by x87** [comment]
### Opcode and Command name
![04](https://user-images.githubusercontent.com/5698288/113787641-68949580-9709-11eb-8995-d55cb9a02a09.PNG)

Opcodes (command id) and command name are two **mandatory** elements that should always exist. Both are **unique** within the given extension. Different extensions may have duplicated command ids or names.

Command Id must have 4 hexadecimal characters and start with `0`. 
Command name must be in SCREAMING_SNAKE_CASE where all letters are in their capital form and words separated with `_`. No extra characters allowed. Editor replaces all whitespace characters and `-` with `_`

**Comment by x87** [comment]
### Extension Picker

![05](https://user-images.githubusercontent.com/5698288/113788017-261f8880-970a-11eb-9f1f-2518ba64be27.PNG)

The extension is a **mandatory** field. Each command must fall in either one of the available extensions. If no extension name provided, the editor uses the default extension named `default`.

Clicking the arrow button opens the name picker
![06](https://user-images.githubusercontent.com/5698288/113788155-6979f700-970a-11eb-8d96-cd6db0b1de13.PNG)

Entering a name of the extension that does not exist yet creates it. Extension names must be **unique** and be known to Sanny Builder. Refer to this list https://docs.sannybuilder.com/edit-modes/extensions#extensions-list for allowed names.

**Comment by x87** [comment]
### Class Picker

![image](https://user-images.githubusercontent.com/5698288/115168426-dd13f080-a088-11eb-9e20-2665a429c0cc.png)

You can pick the class name by clicking the button left to the input field. In the class name picker you can either choose one of the already available names or enter a new name in the bottom field (same as in the [Extension name picker](https://github.com/orgs/sannybuilder/teams/editors/discussions/2/comments/3)). Only letters and numbers are allowed. The name should be in UpperCamelCase. The editor automatically converts the input string into the required format.

**Comment by x87** [comment]
### Class Member Name

![image](https://user-images.githubusercontent.com/5698288/115168465-fddc4600-a088-11eb-9b21-fe7a95236156.png)

Only letters and numbers are allowed. The name should be in UpperCamelCase. The editor automatically converts the input string into the required format.

**Comment by x87** [comment]
### Short Description

![image](https://user-images.githubusercontent.com/5698288/115168496-1b111480-a089-11eb-8204-a5f5f547b2f7.png)

Short description is a single sentence (without the ending period) that tries to highlight the meaning of this command. Do not try to describe everything here. For consistency with other commands start the description with a verb (e.g. `Sets`, `Makes`, `Removes`, etc.). Start conditional commands description with `Returns true if` or `Returns true when`.

**Comment by x87** [comment]
### Input Parameters

![image](https://user-images.githubusercontent.com/5698288/115168529-4ac01c80-a089-11eb-9427-33bfd07ed80c.png)

**Reply by x87** [reply]
You can reorder rows by dragging a number cell

**Comment by x87** [comment]
### Output Parameters

![image](https://user-images.githubusercontent.com/5698288/115168552-5c092900-a089-11eb-9dbe-79b4ba0b8d65.png)

**Reply by x87** [reply]
You can reorder rows by dragging a number cell

**Comment by x87** [comment]
### Attributes

![image](https://user-images.githubusercontent.com/5698288/115168603-91157b80-a089-11eb-8d35-8d844787e94a.png)

Attributes serve to provide extra information about command specifics. The list of attributes is finite. Each command may have one or many attributes, or do not have any.
* `is_branch`: **This command is used to branch the code. Such commands commonly have a label parameter**.
* `is_condition`: **This command can be used in conditional statements, e.g. be placed between IF..THEN**. 
* `is_constructor`: **This command is used to create a new in-game entity, e.g. a vehicle or a pickup.**
* `is_destructor`: **This command is used to delete the in-game entity previously created by an `is_constructor` command**
* `is_keyword`: **This command is used as a single keyword followed by arguments. Keywords are part of the language constructs and directly affect the code execution**
* `is_nop`: **This command is a no-operation, just eats the arguments**
* `is_unsupported`: **This command is unsupported in the given game and its usage is forbidden**
* `is_overload`: **This command has multiple variations for different types of arguments**
* `is_segment`: **This command is used to separate segments in SCM header. As of now it's only the command `jump` (opcode 0002)**
* `is_static`: **This command operates on a static property or in-game entity that can not be constructed dynamically**
* `is_variadic`: **This command has variadic number of arguments, see `start_new_script` for example**

While a command may have multiple attributes, some of them are mutually exclusive. For example, the command can not be `is_constructor` and `is_destructor` at the same time. The editor will display an error if the current combination of attributes is forbidden.

![image](https://user-images.githubusercontent.com/5698288/115168942-a9d26100-a08a-11eb-9250-547b023f788e.png)

**Comment by x87** [comment]
### Code Snippet

![image](https://user-images.githubusercontent.com/5698288/115168641-a5597880-a089-11eb-82f4-fac611ed83fb.png)

**Comment by x87** [comment]
### Control Elements

![image](https://user-images.githubusercontent.com/5698288/115168663-b609ee80-a089-11eb-93cf-69ce6e27c55c.png)

If there are any unsaved changes, the editor displays the `Reset all changes` link. Clicking it will revert the command into the initial state.

`Close` button reverts all changes made to the point and closes the editor.

`Save` button stores the changes and makes them available for a commit to GitHub, then closes the editor.

**Comment by GuitaristForEver** [comment]
**Update:** PR #7375 has been updated to implement the formula-based approach (Option C) per your guidance.

### Key Changes:
- New `triggerScoped` fallback behavior that passes `nil` to formula for failed triggers
- Uses nil-coalescing operator: `primary ?? secondary ?? 5`
- Global `failureThreshold` at ScaledObject level (not per-trigger)
- Integrates with PR #7346's `TriggersActivity` for observability
- Leverages existing `Health` status tracking

Thanks for the clear direction! This approach is much cleaner than the initial per-trigger configuration.
